desc/objectives/_free_boundary.py

```diff
@@ -0,0 +1,945 @@
+        else:
+            return jnp.concatenate([Bn_err, Bsq_err])
+
+    def print_value(self, *args, **kwargs):
+        """Print the value of the objective."""
+        # this objective is really 3 residuals concatenated so its helpful to print
+        # them individually
+        f = self.compute_unscaled(*args, **kwargs)
+        # try to do weighted mean if possible
+        constants = kwargs.get("constants", self.constants)
+        if constants is None:
+            w = jnp.ones_like(f)
+        else:
+            w = constants["quad_weights"]
+
+        abserr = jnp.all(self.target == 0)
+
+        def _print(fmt, fmax, fmin, fmean, norm, units):
+
+            print(
+                "Maximum " + ("absolute " if abserr else "") + fmt.format(fmax) + units
+            )
+            print(
+                "Minimum " + ("absolute " if abserr else "") + fmt.format(fmin) + units
+            )
+            print(
+                "Average " + ("absolute " if abserr else "") + fmt.format(fmean) + units
+            )
+
+            if self._normalize and units != "(dimensionless)":
+                print(
+                    "Maximum "
+                    + ("absolute " if abserr else "")
+                    + fmt.format(fmax / norm)
+                    + "(normalized)"
+                )
+                print(
+                    "Minimum "
+                    + ("absolute " if abserr else "")
+                    + fmt.format(fmin / norm)
+                    + "(normalized)"
+                )
+                print(
+                    "Average "
+                    + ("absolute " if abserr else "")
+                    + fmt.format(fmean / norm)
+                    + "(normalized)"
+                )
+
+        formats = [
+            "Boundary normal field error: {:10.3e} ",
+            "Boundary magnetic pressure error: {:10.3e} ",
+            "Boundary field jump error: {:10.3e} ",
+        ]
+        units = ["(T)", "(T^2)", "(T)"]
+        if self._sheet_current:
+            nn = f.size // 3
+            norms = [
+                self.normalization[0],
+                self.normalization[nn],
+                self.normalization[-1],
+            ]
+        else:
+            nn = f.size // 2
+            formats = formats[:-1]
+            units = units[:-1]
+            norms = [self.normalization[0], self.normalization[nn]]
+        for i, (fmt, norm, unit) in enumerate(zip(formats, norms, units)):
+            fi = f[i * nn : (i + 1) * nn]
+            # target == 0 probably indicates f is some sort of error metric,
+            # mean abs makes more sense than mean
+            fi = jnp.abs(fi) if abserr else fi
+            wi = w[i * nn : (i + 1) * nn]
+            fmax = jnp.max(fi)
+            fmin = jnp.min(fi)
+            fmean = jnp.mean(fi * wi) / jnp.mean(wi)
+            _print(fmt, fmax, fmin, fmean, norm, unit)
+
+
+class BoundaryErrorNESTOR(_Objective):
```

Comment:
We don't have any examples of using this objective, right? I think I only see it covered in a single unit test. That might be fine if we don't really want people to use it, but I want to understand how well it is tested.

Reply:
Yeah its both slower and less accurate so it's not really used. It's also somewhat memory intensive so it's hard to come up with a real example test that can run on CI